gerrit_changes_to_github.py

```diff
@@ -271,6 +183,82 @@ def is_repository_usable(args):
     return got_gerrit and got_target
 
 
+def get_changes_from_event_logs(args, gerrit_password, ghpa_token):
+    """Go through all events since the latest timestamp"""
+    try:
+        latest_timestamp = int(os.getenv("LAST_TIMESTAMP"))
+    except Exception as e:
+        print(f"Error reading last timestamp: {e}")
+
+    latest_ts_as_date = datetime.fromtimestamp(
+        latest_timestamp, tz=timezone.utc
+    ).strftime("%Y-%m-%d %H:%M:%S")
+    log.info(f"Using last timestamp (UTC): {latest_ts_as_date}")
+
+    if (
+        events := requests.get(
+            f"https://review.spdk.io/a/plugins/"
+            f"events-log/events/?t1={latest_ts_as_date}",
+            auth=("spdk-community-ci-samsung", gerrit_password),
+        )
+    ).status_code != 200:
+
+        log.error(f"Failed to retrieve events. HTTP Status Code: {events.status_code}")
+        return None
+
+    events = events.text.replace("}{", "},{").replace("}\n{", "},{")
+    events = json.loads(f"[{events}]")
+
+    refs = []
+    for event in events:
+        event_ts = event.get("eventCreatedOn")
+        latest_timestamp = event_ts if event_ts > latest_timestamp else latest_timestamp
+
+        # Get new changes and new revisions on existing changes
+        if event.get("type") == "patchset-created" and "RFC" not in event.get(
+            "change"
+        ).get("commitMessage"):
+            refs.append(event.get("patchSet").get("ref"))
+        # Get changes marked "false positive"
+        elif event.get("type") == "comment-added" and "false positive" in event.get(
+            "comment"
+        ):
+            refs.append(event.get("patchSet").get("ref"))
+
+    # Be verbose; it makes it easier to understand what is going on by inspecting the
+    # output of the command in a CI log and even when running it manually
+    for i in range(len(refs)):
+        log.info(f"Got ref({refs[i]}) via REST API")
+    log.info(f"A total of {len(refs)} changes retrieved.")
+
+    if len(refs) >= args.limit:
+        log.info(
+            f"There is likely more to process, as n equals the limit({args.limit})"
+        )
+
+    latest_ts_as_date = datetime.fromtimestamp(
+        latest_timestamp, tz=timezone.utc
+    ).strftime("%Y-%m-%d %H:%M:%S")
+    log.info(f"Saving last timestamp (UTC): {latest_ts_as_date}")
+
+    # Update LATEST_TIMESTAMP
+    url = (
+        "https://api.github.com/repos/spdk-community-ci/dispatcher"
+        "/actions/variables/LAST_TIMESTAMP"
+    )
+    headers = {
+        "Accept": "application/vnd.github.v3+json",
+        "Authorization": f"token {ghpa_token}",
+    }
+    data = {"value": f"{latest_timestamp}"}
```

Comment:
let's say the last event has a timestamp of X

if next time we request events with t1=X, will it return that same event again?

or do we need to save X+1 as the timestamp?

this is kind of tricky, we also need to avoid corner case where 2 events happen at timestamp X, but we happen to query right between the two events - we get A but not B, if we save off X + 1 as the last timestamp, then we'll never process B

what does the JSON for these events look like? is there any kind of event ID?

Reply:
Resolved during our meeting: To avoid getting a fraction of a second of events (ex: multiple events have the same creation time and we make a query in the middle of a second, getting half the events for that second), we change the query to have t1=start_time and t2=now-1 to get all the events from t1 to t2.